File: VB/E4924/Models/CustomAttribute.vb
Imports System.ComponentModel.DataAnnotations

Public Class CustomAttribute
    Inherits ValidationAttribute
    Implements IClientValidatable

    Public Property MinAge As Integer

    Protected Overrides Function IsValid(ByVal value As Object, ByVal validationContext As ValidationContext) As ValidationResult
        Dim p As Person = TryCast(validationContext.ObjectInstance, Person)
        If p Is Nothing Then Return New ValidationResult("Internal  error")
        If Not p.CheckAge Then Return ValidationResult.Success
        Dim ts As TimeSpan = DateTime.Now - p.BirthDate

        If ts.TotalDays / 365 < MinAge Then
            Return New ValidationResult("Age must be more than " & MinAge.ToString())
        Else
            Return ValidationResult.Success
        End If
    End Function

    Public Function GetClientValidationRules(metadata As ModelMetadata, context As ControllerContext) As IEnumerable(Of ModelClientValidationRule) Implements IClientValidatable.GetClientValidationRules
        Dim l As List(Of ModelClientValidationRule) = New List(Of ModelClientValidationRule)
        Dim s As String = "Age must be more than " & MinAge.ToString()
        Dim r As ModelClientValidationRangeDateRule = New ModelClientValidationRangeDateRule(s, MinAge)
        l.Add(r)
        Return l
    End Function

    Public Class ModelClientValidationRangeDateRule
        Inherits ModelClientValidationRule

        Public Sub New(ByVal errorMessage As String, ByVal minAge As Integer)
            Me.ErrorMessage = errorMessage
            Me.ValidationType = "rangedate"
            Me.ValidationParameters("minage") = minAge
        End Sub
    End Class
End Class

File: CS/E4924/Models/CustomAttribute.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace E4924.Models
{
    public class CustomAttribute : ValidationAttribute, IClientValidatable
    {
        public int MinAge { set; get; }

        protected override ValidationResult IsValid(object value, ValidationContext validationContext)
        {
            Person p = validationContext.ObjectInstance as Person;
            if (p == null)
                return new ValidationResult("Internal  error");
            if (!p.CheckAge)
                return ValidationResult.Success;
            TimeSpan ts = DateTime.Now - p.BirthDate;
            if (ts.TotalDays / 365 < MinAge)
                return new ValidationResult("Age must be more than " + MinAge.ToString());
            else
                return ValidationResult.Success;
        }

        public IEnumerable<ModelClientValidationRule> GetClientValidationRules(ModelMetadata metadata, ControllerContext context)
        {
            yield return new ModelClientValidationRangeDateRule("Age must be more than " + MinAge.ToString(), MinAge);
        }

        public class ModelClientValidationRangeDateRule : ModelClientValidationRule
        {
            public ModelClientValidationRangeDateRule(string errorMessage, int minAge)
            {
                ErrorMessage = errorMessage;
                ValidationType = "rangedate";
                ValidationParameters["minage"] = minAge;
            }
        }
    }
}